Sawtooth RFC for Docker Build and Execution in a Proxy Environment

danintel opened the pull request:
Signed-off-by: danintel <daniel.anderson@intel.com>

This PR submits the Docker proxy RFC after final reviews and modifications by everyone interested with the Google Doc:

https://docs.google.com/document/d/1Xux_qbx8edu5_bWYZsnfr7p5JoC3ZpPCaoDhhykChno/edit

agunde406 commented on text/0045-docker-proxy.md:
```diff
@@ -0,0 +1,115 @@
+# Summary
+[summary]: #summary
+
+Sawtooth has containerized build and test processes for core and all components. Running Docker containers behind a network proxy requires special handling so that the containers can find external routes to download packages and build dependencies, for example. There are a number of ways to configure these environment for proxy support. It is desirable that we adopt a single consistent approach a ...
```
Please wrap the lines to 80 characters

danintel replied:
Lines wrapped.

arsulegai commented on text/0045-docker-proxy.md:
````diff
@@ -0,0 +1,148 @@
+ | apt-key add -
+```
+
+All Docker Compose files must, pass `http_proxy` , `https_proxy` , and `no_proxy` through to each
````
This depends on the docker version. Maybe we can mention that Sawtooth assumes user is on latest version and in version 18.03 onwards this explicit passing of env is not required.

danintel replied:
I didn't know that. I will verify the *_proxy variables are passed to Docker and update the document.

danintel replied:
I tried a simple test that removed the `*_proxy` variables from a Docker file and the application now fails.  I used my CookieJar application with docker 18.09.6 and docker-compose 1.21.0.  Here is the diff:

```
$ git diff
diff --git a/docker-compose.yaml b/docker-compose.yaml
index 934b1ee..8ac2031 100644
--- a/docker-compose.yaml
+++ b/docker-compose.yaml
@@ -39,10 +39,6 @@ services:
         - http_proxy
         - https_proxy
         - no_proxy
-    environment:
-      - 'http_proxy=${http_proxy}'
-      - 'https_proxy=${https_proxy}'
-      - 'no_proxy=rest-api,validator,${no_proxy}'
     volumes:
       - '.:/project/cookiejar/'
     depends_on:
```

The appliation worked before the change. After rebuilding the docker container with the above change, this is the result:

```
$ sudo docker exec -it cookiejar-cl
ient bash
[sudo] password for dano: 
root@e18112d57b62:/project/cookiejar/pyclient# cookiejar.py bake 100
URL to send to REST API is http://rest-api:8008/batches
[18:47:26 INFO    connectionpool] Starting new HTTP connection (1): proxy-jf.intel.com
[18:47:26 DEBUG   connectionpool] "POST http://rest-api:8008/batches HTTP/1.1" 404 654
Traceback (most recent call last):
  File "/project/cookiejar/pyclient/cookiejar_client.py", line 137, in _send_to_rest_api
    result.status_code, result.reason))
Exception: Error 404: Not Found
```

docker and docker-compose version:
```
 $ docker version
Client:
 Version:           18.09.6
 API version:       1.39
 Go version:        go1.10.8
 Git commit:        481bc77
 Built:             Sat May  4 02:35:57 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.6
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.8
  Git commit:       481bc77
  Built:            Sat May  4 01:59:36 2019
  OS/Arch:          linux/amd64
  Experimental:     false
daniela1-test1 sawtooth-cookiejar (docker) $ docker compose version
docker: 'compose' is not a docker command.
See 'docker --help'
daniela1-test1 sawtooth-cookiejar (docker) $ docker-compose version
docker-compose version 1.21.0, build 5920eb0
docker-py version: 3.2.1
CPython version: 3.6.5
OpenSSL version: OpenSSL 1.0.1t  3 May 2016
```

arsulegai replied:
There are two things here:
1. If you pruned all docker images and rebuilt them by removing `args` under the `build` then `config.json` worked as expected. Proxy env variables were passed by docker during build time.
2. These env are also passed during container runtime, hence you saw see the error while executing here. By adding `rest-api` to the `no_proxy` list in `config.json` this issue would be resolved.

We're on same page, how do we handle `no_proxy` is the next question.

danintel replied:
Sorry, I still do not understand what's being proposed here. This `config.json` file. Does it go in the host? Where? `/etc/docker/config.json` ? Or does it go in each of the Docker containers? If it's the latter, that is much less convenient than just setting environment variables.

Also what environment variables are now automatically passed to containers? Is it `http_proxy` , `https_proxy` , and `no_proxy` ? If these are passed, why is a config.json needed?

If `config.json` (is it `/etc/docker/config.json` ?), can you give me an example?

BTW, I removed the `args` and rebuilt and it still did not work.

arsulegai replied:
Ah! I assumed you were aware of these changes from your previous comment. Let me brief it now, `config.json` is a file to be created on the client machine where we want to run the containers. Create this file under `~/.docker/` directory. This file will have the `****_proxy` and other env variables which are to be passed by the docker during build and running the container.

The `config.json` file looks like
```
{
    "proxies":
    {
        "default":
        {
            "httpProxy": "http://proxy-server-domain:<port>",
            "httpsProxy": "http://proxy-server-domain:<port>",
            "noProxy": "*.my-company.com,localhost,rest-api*",
            "hkpProxy": "http://proxy-server-domain:<port>"
        }
    }
}
```

Please let me know if adding this file helped you

danintel commented:
I updated the RFC to use `~/.docker/config.json` after testing it and getting it working.
The tricky part was making sure the Docker internal networks were included in `noProxy`.

dcmiddle commented on text/0045-docker-proxy.md:
````diff
@@ -0,0 +1,164 @@
+ --recv-keys 8AA7AF1F1091A5FD
+```
+
+Use the following, which works in proxy (with http_proxy) and non-proxy environments:
````
strike the parenthetic, based on recommendation to use the config.json file.

danintel replied:
Fixed.  Please review and approve.

vaporos commented:
I like where where this has gone--less proxy support within the docker files themselves, but specific information on how to setup the proxy configuration.

The majority of the RFC contains explanation on how to use docker, instead of changes we need to make with Sawtooth. The changes are to remove what we currently have for proxy support. Do we still want to move forward with the RFC (which would now be an RFC to not have explicit proxy support), or should we just move forward with PRs to simplify the files and document how to setup proxy (which could be derived from the text of this RFC).

danintel commented:
Yes, I want to move forward as a RFC. The one part where (minor) changes are still needed is for key servers. The rest is good for informational reasons.

Follow-up would be removing unneeded proxy configuration/passthrough in Docker files, as they are found.

Dan

From: Shawn T Amundson <notifications@github.com>
Sent: Tuesday, July 2, 2019 10:04 AM
To: hyperledger/sawtooth-rfcs <sawtooth-rfcs@noreply.github.com>
Cc: Anderson, Daniel <daniel.anderson@intel.com>; Author <author@noreply.github.com>
Subject: Re: [hyperledger/sawtooth-rfcs] Sawtooth RFC for Docker Build and Execution in a Proxy Environment (#45)


I like where where this has gone--less proxy support within the docker files themselves, but specific information on how to setup the proxy configuration.

The majority of the RFC contains explanation on how to use docker, instead of changes we need to make with Sawtooth. The changes are to remove what we currently have for proxy support. Do we still want to move forward with the RFC (which would now be an RFC to not have explicit proxy support), or should we just move forward with PRs to simplify the files and document how to setup proxy (which could be derived from the text of this RFC).

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/hyperledger/sawtooth-rfcs/pull/45?email_source=notifications&email_token=AH26URIFM25GYCFEGBJRPP3P5OC7ZA5CNFSM4HZKS2IKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZB5YMI#issuecomment-507763761>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AH26URJHOY4FXRMQH7VMBG3P5OC7ZANCNFSM4HZKS2IA>.

dcmiddle commented:
> I like where where this has gone--less proxy support within the docker files themselves, but specific information on how to setup the proxy configuration.
> 
> The majority of the RFC contains explanation on how to use docker, instead of changes we need to make with Sawtooth. The changes are to remove what we currently have for proxy support. Do we still want to move forward with the RFC (which would now be an RFC to not have explicit proxy support), or should we just move forward with PRs to simplify the files and document how to setup proxy (which could be derived from the text of this RFC)?

That seems like a valid reason for an RFC.. here's the architectural pattern we've adopted across all of the components and repos. We can refer to it to avoid inadvertent bikeshedding. It may be a little lighter weight than some of the other RFCs, but all the better if we can move things along expediently.

danintel commented:
Can I get approvals here for this Proxy RFC now that the issues are resolved? I see one approval from Dan M.

arsulegai commented on text/0045-docker-proxy.md:
```diff
@@ -0,0 +1,167 @@
+Starting with Ubuntu 18 (Bionic), `gpg` , invoked through `apt-key adv` ,
+now ignores `http_proxy` and `https_proxy` environment variables, and ignores
+the command line options
+`apt-key adv --keyserver-options http-proxy=<proxy server url>`
```
Ubuntu 18 doesn't ignore command line options, however it would be too much of a code in every Dockerfile if we choose this way.

danintel replied:
The current version of `gpg` does ignore the `http-proxy` and `https-proxy` command line options. I found this out from testing on Bionic and confirmed it with multiple documents on the web. This document, for example, says:

```http-proxy=value
...
These options have no more function since GnuPG 2.1```

https://www.gnupg.org/documentation/manuals/gnupg/GPG-Configuration-Options.html

arsulegai replied:
Ah! I am not sure what version of gpg gets installed. These CLI worked for me. Thanks for pointing it out.

rbuysse replied:
Please include the gnupg link above in the RFC.

danintel replied:
Added a link.

arsulegai commented on text/0045-docker-proxy.md:
```diff
@@ -0,0 +1,167 @@
+The current alternative used is hand-modifying Docker scripts each time a
+proxy is required. This is time consuming, expensive, and unnecessary.
+
+An earlier alternative of using `apt-key adv` and passing proxy variables on
```
This paragraph needs changes. Xenial did not require `--keyserver-options` explicitly passing proxy env. `config.json` in `~/.docker/` is sufficient for Xenial. The concern is however with Bionic, where all Sawtooth Dockerfiles are moved to.

danintel replied:
I do not mention that `--keyserver-options` is required (one can also use environment variables, for example, in Xenial).. I do mention that `--keyserver-options` was used on line 36 (not lines 140-143).

It is good to know `config.json` is also supported in Xenial, as it makes this proposal to use `config.jason` stronger.

rbuysse commented on text/0045-docker-proxy.md:
```diff
@@ -0,0 +1,167 @@
+* Have Docker and Docker Compose files and shell scripts that work in explicit
+  proxy, transparent proxy, and non-proxy environments
+* Have a uniform method to access proxy servers across Hyperledger Sawtooth
+  projects * Minimize Docker and script clutter with unnecessary tests and
```
Is "* Minimize Docker..." meant to be its own bullet point?

danintel replied:
Yes. Looks like a copy paste error from the original Google Doc discussion.

danintel replied:
Fixed.

rberg2 commented on text/0045-docker-proxy.md:
````diff
@@ -0,0 +1,167 @@
+
+```
+RUN \
+ apt-get update && apt-get install -yq curl && curl -sSL \
````
I would like to see `--retry 3` or some other number in the curl arguments since we sometimes see intermittent failures due to downloading keys from the Ubuntu keyserver. Currently we have a fallback to the p80 pool if this server fails. Thoughts?

arsulegai replied:
Sounds good. How about having a curl try on fallback to the p80 pool if first 3 attempts fail?

danintel replied:
The original attempt was to avoid a lot of boilerplate code. If adding boilerplate code is now OK with everyone, I can do that.

arsulegai replied:
No comments

danintel replied:
I added `--retry 3` and the  `p80.pool.sks-keyservers.net` keyserver. Both of these are not related to proxies, but it is also good practice.

rbuysse commented on text/0045-docker-proxy.md:
```diff
@@ -0,0 +1,167 @@
+variables has the disadvantage of requiring these variables to be explicitly
+passed through the `args:` and `environment:` sections for each container in a `docker-compose.yaml` file.
+
+A forth alternative, setting configuration for newer versions of gpg (in
```
s/forth/fourth

danintel replied:
Fixed

rbuysse commented on text/0045-docker-proxy.md:
```diff
@@ -0,0 +1,167 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
```
I'd prefer that this link to the official docker documentation rather than try to duplicate it in case the process changes in the future.

danintel replied:
I added a link to the official Docker document. It does not cover the special case of key servers though.

rbuysse replied:
Can you move the docker documentation link to this section instead? It's relevant in this section, but not to the one below.

danintel replied:
Yes.

rbuysse commented:
I am still skeptical that this is appropriate material for an RFC but I approve of the direction this is taking and I will not disagree if the consensus is to codify this here.

danintel commented:
I addressed all comments above for the proxy RFC. Let me know if there are more issues or questions.

rbuysse commented on text/0045-docker-proxy.md:
````diff
@@ -0,0 +1,176 @@
+two keyservers for robustness:
+
+```
+RUN \
````
Since this RFC is going to become the reference document, I'd like to see this section be more concise and readable.

If we're adding retries, I think we can dispense with the backup keyserver.

Also, this doesn't work unless gnupg2 is installed.

Suggestion:
```
RUN apt-get update \
 && apt-get install -yq \
 curl \
 gnupg2

RUN curl --retry 3 -sSL \
    'https://keyserver.ubuntu.com/pks/lookup?op=get&search=0x8AA7AF1F1091A5FD' \
    | apt-key add -
```

rberg2 replied:
This is what I had in mind with my earlier comment about using a curl retry.

danintel replied:
I removed the alternate keyserver as suggested by others and kept the `--retry 3`
Retry logic and alternate keyservers are not really a proxy issue.

danintel replied:
Package gnupg2 is not needed and it works for me without it being installed.  I only see documentation in the package:

```
dpkg-query -L gnupg
/.
/usr
/usr/share
/usr/share/doc
/usr/share/doc/gnupg
/usr/share/doc/gnupg/DETAILS.gz
/usr/share/doc/gnupg/FAQ
/usr/share/doc/gnupg/HACKING.gz
/usr/share/doc/gnupg/OpenPGP.gz
/usr/share/doc/gnupg/README.Debian
/usr/share/doc/gnupg/README.gz
/usr/share/doc/gnupg/copyright
/usr/share/info
/usr/share/info/gnupg-card-architecture.png
/usr/share/info/gnupg-module-overview.png
/usr/share/info/gnupg.info-1.gz
/usr/share/info/gnupg.info-2.gz
/usr/share/info/gnupg.info.gz
/usr/share/doc/gnupg/NEWS.Debian.gz
/usr/share/doc/gnupg/NEWS.gz
/usr/share/doc/gnupg/THANKS.gz
/usr/share/doc/gnupg/TODO
/usr/share/doc/gnupg/changelog.Debian.gz
```

danintel replied:
After removing packages `gnupg` and documentation package `gnupg2`, both `apt-key adv` and `apt-key add` both still work.  See example below.

```
$ sudo apt remove gnupg gnupg2
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following packages will be REMOVED:
  gnupg gnupg2
0 upgraded, 0 newly installed, 2 to remove and 0 not upgraded.
After this operation, 423 kB disk space will be freed.
Do you want to continue? [Y/n] y
(Reading database ... 184249 files and directories currently installed.)
Removing gnupg2 (2.2.4-1ubuntu1.2) ...
Removing gnupg (2.2.4-1ubuntu1.2) ...
Processing triggers for man-db (2.8.3-2ubuntu0.1) ...
$ sudo apt-key del  44FC67F19B2466EA                              
OK
$ apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys 44FC67F19B2466EA
Executing: /tmp/dano/apt-key-gpghome.yCD8U38W9A/gpg.1.sh --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys 44FC67F19B2466EA
gpg: key 44FC67F19B2466EA: public key "Hyperledger Sawtooth Nightly Signing Key <distributedledger@intel.com>" imported
gpg: Total number processed: 1
gpg:               imported: 1
gpg: no writable keyring found: Not found
gpg: error reading '[stdin]': General error
gpg: import from '[stdin]' failed: General error
gpg: Total number processed: 0
$ sudo apt-key del  44FC67F19B2466EA                              
OK
$ curl --retry 3 -sSL 'https://keyserver.ubuntu.com/pks/lookup?op=get&search=0x44FC67F19B2466EA' | sudo apt-key add -                
OK
```

danintel commented:
I addressed new issues raised in the current round of comments.  So please take a look here.

rbuysse commented:
Can you double check my comments here? https://github.com/hyperledger/sawtooth-rfcs/pull/45#discussion_r307814343

Looks like you removed the backup keyserver but my other concerns weren't addressed.

danintel commented:
@rbysse I checked all your comments and made some changes.  If there are any more issues, please let me know which ones.  Here is a list:

* _Please include the gnupg link above in the RFC._
Added a link around line 42.

* _I'd prefer that this link to the official docker documentation rather than try to duplicate it in case the process changes in the future._
I deleted the text and replaced it with a link under "Guide-level explanation"
around line 62.

* _Can you move the docker documentation link to this section instead? It's relevant in this section, but not to the one below._
Same as above.

* _Is "* Minimize Docker..." meant to be its own bullet point?_
Yes.  Now a separate line on line 52.

* _s/forth/fourth_
Fixed at line 124.
```

* _I am still skeptical that this is appropriate material for an RFC but I approve of the direction this is taking and I will not disagree if the consensus is to codify this here._
OK. The docker-compose/apt-key adv/gpg interaction isn't really documented properly.

*_Since this RFC is going to become the reference document, I'd like to see this section be more concise and readable.  If we're adding retries, I think we can dispense with the backup keyserver. RUN apt-get update && apt-get install -yq curl gnupg2
RUN curl --retry 3 -sSL 'https://keyserver.ubuntu.com/pks/lookup?op=get&search=0x8AA7AF1F1091A5FD' | apt-key add -_

I added the backup keyserver from an earlier suggestion. I removed it again and now there is just the `--retry`

*_Also, this doesn't work unless gnupg2 is installed._
Package gnupg2 is just documentation under /usr/share/doc and /usr/share/info
See a previous comment for gnupg2 package contents.  Perhaps you mean gnupg?  But, I removed both gnupg and gnupg2 and was able to install keys with `apt-key add` and `apt-key add`.

rbuysse commented:
I'm not sure how you're testing this, but the example code you have in the RFC doesn't work in docker.

Trying to build the following dockerfile:

```
FROM ubuntu:bionic

RUN \
 apt-get update && apt-get install -yq curl && curl --retry 3 -sSL \
    'http://keyserver.ubuntu.com/pks/lookup?op=get&search=0x8AA7AF1F1091A5FD' \
    | apt-key add -
```

Exits with error:
```
E: gnupg, gnupg2 and gnupg1 do not seem to be installed, but one of them is required for this operation
(23) Failed writing body
The command '/bin/sh -c apt-get update && apt-get install -yq curl && curl --retry 3 -sSL     'http://keyserver.ubuntu.com/pks/lookup?op=get&search=0x8AA7AF1F1091A5FD'     | apt-key add -' returned a non-zero code: 255
```

The suggestion I provided is more readable, breaking up the two operations into logical steps and it works.

danintel commented:
@rbuysse I have no problem running Sawtooth and apt-get and apt-key without explicitly installing gnupg/gnupg2/gnupg1.  I just tried it on Ubuntu 18.04 LTS, docker-compose 1.24.1.

I also notice that none of the sawtooth-core (and no sawtooth-* repo I checked) ever installs gnupg2 or gnupg1.  Repo sawtooth-core installs gnupg, but that is documentation only, no binary. 

So if package gnupg2 (or gnupg1) is required, why is no Sawtooth Dockerfile installing it?

Also, installing a key is required in both proxy and non-proxy environments and properly should be documented separately. I also notice that the Sawtooth documentation does not mention installing gnupg2:
https://sawtooth.hyperledger.org/docs/core/releases/latest/sysadmin_guide/installation.html

But there is something different about your environment from all others, so I am adding package `gnupg2` to the list. However, this (apt-key dependencies) is really a non-proxy issue.

I updated the RFC so please have a look and approve it. Thanks

peterschwarz commented:
What happens when your run a Docker file containing just this:

```Dockerfile
FROM ubuntu:bionic

RUN \
 apt-get update && apt-get install -yq curl && curl --retry 3 -sSL \
    'http://keyserver.ubuntu.com/pks/lookup?op=get&search=0x8AA7AF1F1091A5FD' \
    | apt-key add -
```
as @rbuysse suggested?

rberg2 commented:
Hi Dan, Gnupg does need to installed for secure apt to work or apt won't be able to verify the signed package. Since Ubuntu Bionic gnupg is not installed by default in the Docker images. Gnupg was added to sawtooth-core [here](https://github.com/hyperledger/sawtooth-core/commit/c23f330a2520e625b58cbeda446f230fca41246e). The gnupg package depends on gnupg-utils and gpg 2 among others and installing it results in a complete installation of gnupg.

danintel commented:
@peterschwarz  It fails, but works when I add `gnupg2`
My guess is the other Docker instances are using pre-built images with gnupg2 previously installed.

In any case, I already updated the docs to specify `gnupg2`

The existing Sawtooth documentation should also be updated to mention that  one of `gnupg2/gnupg1/gnupg` is required. That is not done anywhere.

Again this is not a proxy issue and I already updated this RFC.

danintel commented:
@rberg2 The PR has no documention updates. So the Sawtooth user/administrator is never told to install gnupg2 (or gnupg or gnupg1).

In any case this RFC mentions gnupg2 and is now ready to be approved.

danintel commented:
I created a separate PR in sawtooth-core to add missing documentation about installing `gnupg2` when installing Sawtooth. Please see https://github.com/hyperledger/sawtooth-core/pull/2222

rbuysse commented:
Also I'd still like to see the docker example code cleaned up as mentioned in this comment: https://github.com/hyperledger/sawtooth-rfcs/pull/45#discussion_r307814343

danintel commented:
@rbuysse Reading between the lines, I assume the important part of your comment is having two `RUN` stanzas, not one.  I updated the RFC to have two `RUN` stanzas.

danintel commented:
@rberg2 I removed the comment about two keyservers.

vaporos commented:
We should either be clear that this applies only to the specific Ubuntu versions, or we should make it more principle-based so that it would apply to CentOS/RHEL as well.

danintel commented:
@vaporos Ubuntu is specifically mentioned under "Summary". It does not apply to CentOS/RHEL as they do not use APT, at least in general.

I could emphasize the restriction more.
